Change the diagnosticInfo.Id to an DataSetWriter/Endpoint

dacolgit opened the pull request:
Added endpoint info in diagnostic model
added diagnosticApiModel
fix doc

dacolgit committed:
Added endpoint info in diagnostic model
added diagnosticApiModel
fix doc

dacolgit committed:
Merge branch 'main' into feature/dacol-IdData

dacolgit committed:
fix e2e test

dacolgit committed:
fix

dacolgit committed:
remove empty lines

karok2m commented on api/src/Microsoft.Azure.IIoT.OpcUa.Api.Publisher/src/Models/DiagnosticInfoApiModel.cs:
```diff
@@ -0,0 +1,132 @@
+        /// <summary> EndpointInfo /// </summary>
+        [DataMember(Name = "PublishedNodesEntryModel", Order = 0,
+            EmitDefaultValue = false)]
+        public PublishedNodesEntryModel EndpointInfo { get; set; }
```
Response API model should be using `PublishNodesEndpointApiModel` instead of `PublishedNodesEntryModel`. `PublishedNodesEntryModel` is the internal class and `PublishNodesEndpointApiModel` is the one intended for external communication.

dacolgit replied:
changed

karok2m commented on components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/StandaloneJobOrchestrator.cs:
```diff
@@ -170,6 +170,24 @@ IJsonSerializer jsonSerializer
+                                var id = _publishedNodesJobConverter.
+                                    ToConnectionModel(entry, _standaloneCliModel).CreateConnectionId();
+
+                                if (_publisherDiagnosticInfo[diagInfo.Id].Id == id) {
```
`_publisherDiagnosticInfo[diagInfo.Id].Id` is the same as just `diagInfo.Id`.

dacolgit replied:
changed

karok2m commented on components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/StandaloneJobOrchestrator.cs:
```diff
@@ -170,6 +170,24 @@ IJsonSerializer jsonSerializer
+                                    ToConnectionModel(entry, _standaloneCliModel).CreateConnectionId();
+
+                                if (_publisherDiagnosticInfo[diagInfo.Id].Id == id) {
+                                    _publisherDiagnosticInfo[id].EndpointInfo = new PublishedNodesEntryModel {
```
The main diagnostic reporting logic is not using `EndpointInfo` at all. We only need it here where we try to find corresponding `PublishedNodesEntryModel` object for each `JobDiagnosticInfoModel` object. So I would suggest to remove `EndpointInfo` field from `JobDiagnosticInfoModel` and instead change `_publisherDiagnosticInfo` to be a dictionary from string `jobId` to `Tuple<JobDiagnosticInfoModel, PublishedNodesEntryModel>`.

This way we can keep the information about corresponding objects and not require cross-project dependencies that were introduced in the PR. In this case the return type from `GetDiagnosticInfoAsync()` would change to `List<Tuple<JobDiagnosticInfoModel, PublishedNodesEntryModel>>` and then the `ToApiModel()` method can take care of transforming those into `DiagnosticInfoApiModel`.

dacolgit replied:
removed `EndpointInfo` from `JobDiagnosticInfoModel`.
I also remove the  cross-project dependencies. 
`_publisherDiagnosticInfo` is now  `Dictionary<string, Tuple<JobDiagnosticInfoModel, PublishedNodesEntryModel>>`.
Using a Dictionary instead of List I was able to keep all the dictionary properties used to manage the JobDiagnosticInfoModel

karok2m commented on e2e-tests/IIoTPlatform-E2E-Tests/Standalone/C_PublishMultipleEndpointsStandaloneDirectMethodTestTheory.cs:
```diff
@@ -234,7 +234,7 @@ IIoTMultipleNodesTestContext context
 
             Assert.Equal((int)HttpStatusCode.OK, responseGetDiagnosticInfo.Status);
-            var diagInfoList = _serializer.Deserialize<List<JobDiagnosticInfoModel>>(responseGetDiagnosticInfo.JsonPayload);
+            var diagInfoList = _serializer.Deserialize<List<DiagnosticInfoApiModel>>(responseGetDiagnosticInfo.JsonPayload);
```
Since response should contain endpoint definition now, we should validate it as well.

dacolgit replied:
added

dacolgit committed:
fix pr comments

dacolgit committed:
removed EndpointInfo from JobDiagnosticInfoModel
removed cross-project dependencies
add E2E test for new endpoint info returned fron GetDiagnosticInfo

karok2m commented on components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/StandaloneJobOrchestrator.cs:
```diff
@@ -169,7 +169,24 @@ IJsonSerializer jsonSerializer
                         if (diagInfo.Id == assignedJob.Value.Job.Id) {
-                            _publisherDiagnosticInfo.AddOrUpdate(assignedJob.Value.Job.Id, diagInfo);
+                            foreach (var entry in _publishedNodesEntries) {
+                                var id = _publishedNodesJobConverter.
```
Now lookin at this again I understand that `jobId` is going to be determined only by the following fields:

* `EndpointUrl`
* `UseSecurity`
* `OpcAuthenticationMode` together with `OpcAuthenticationUsername` and `OpcAuthenticationPassword` 
* `DataSetWriterGroup`

So `DataSetWriterId` and `DataSetPublishingInterval` do not influence `jobId`. This makes sense given that only the above fields are going to define `jobId` and corresponding session details. And fields such as `DataSetWriterId` and `DataSetPublishingInterval` are only going to affect names of subscriptions within the session. What this means is that there might be several entries in `_publishedNodesEntries` list which would result in the same `jobId`. So only taking the first one does not seem to be the correct approach, neither does taking all of them. So maybe we should take only the fields whose values influence the `jobId`, meaning that we exclude `DataSetWriterId` and `DataSetPublishingInterval` from the assignment bellow. Then all endpoint definitions that match on the fields that we take would be part of the diagnostics info. @dacolgit, @cristipogacean, what you guys think about this ?

The other problem that we have here is that we are iterating over `_publishedNodesEntries` now, which means that we have to use `_lockConfig` together with `_lockJobs` instead of only having `_lockJobs` on line `117` above. This is problematic because we introduced `_lockJobs` so that heartbeat reporting can happen faster and not be affected as much by direct method calls. @cristipogacean, is there a way to get the meaningful fields that I've mentioned above from the job definition itself ?

dacolgit replied:
Extended the messageTrigger to export the following info:
- `EndpointUrl
- UseSecurity
- OpcAuthenticationMode together with OpcAuthenticationUsername and OpcAuthenticationPassword
- DataSetWriterGroup`

As discussed the `DataSetWriterId` and 'DataSetPublishingInterval' are removed from the 'EndpointDiagnosticInfo'

cristipogacean committed:
Merge branch 'main' into feature/dacol-IdData

dacolgit committed:
Extended the messageTrigger to export the following info:

`EndpointUrl
UseSecurity
OpcAuthenticationMode together with OpcAuthenticationUsername and OpcAuthenticationPassword
DataSetWriterGroup`
The DataSetWriterId and 'DataSetPublishingInterval' are removed from the 'EndpointDiagnosticInfo'
fix E2E test

dacolgit committed:
Merge branch 'feature/dacol-IdData' of https://github.com/Azure/Industrial-IoT into feature/dacol-IdData

karok2m commented on components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/IPublisherConfigServices.cs:
```diff
@@ -6,6 +6,7 @@
 namespace Microsoft.Azure.IIoT.OpcUa.Edge.Publisher {
     using Microsoft.Azure.IIoT.Agent.Framework.Models;
     using Microsoft.Azure.IIoT.OpcUa.Publisher.Config.Models;
+    using System;
```
This seem to be the only addition to the file. Is it necessary ? My assumption is that it was compiling previously.

dacolgit replied:
removed

karok2m commented on common/src/Microsoft.Azure.IIoT.Agent.Framework/src/Models/JobDiagnosticInfoModel.cs:
```diff
@@ -11,11 +11,23 @@ namespace Microsoft.Azure.IIoT.Agent.Framework.Models {
+        public JobDiagnosticInfoModel() {
+            EndpointInfo = new EndpointDiagnosticModel();
+        }
+
```
We can remove the constructor. `null` for `EndpointDiagnosticModel` does seem like a valid value until it is set.

dacolgit replied:
I added the constructor to avoid null reference exception in `GetDiagnosticInfo` in DataFlowProcessingEngine

dacolgit replied:
just read your comment below. Constructor removed

karok2m commented on components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/DataFlowProcessingEngine.cs:
```diff
@@ -164,6 +163,11 @@ public class DataFlowProcessingEngine : IProcessingEngine, IDisposable {
             double estimatedMsgChunksPerDay = Math.Ceiling(chunkSizeAverage) * sentMessagesPerSec * 60 * 60 * 24;
             var diagnosticInfo = new JobDiagnosticInfoModel();
 
+            diagnosticInfo.EndpointInfo.EndpointUrl = _messageTrigger.EndpointUrl;
```
Here you can first create a `EndpointDiagnosticModel` object, set its fields and then assign it to `diagnosticInfo.EndpointInfo`. That way there is no need to have a special constructor for `JobDiagnosticInfoModel`.

cristipogacean committed:
Merge branch 'main' into feature/dacol-IdData

karok2m committed:
Merge branch 'main' into feature/dacol-IdData

karok2m committed:
Merge branch 'main' into feature/dacol-IdData

dacolgit committed:
fix pr comments

dacolgit committed:
Merge branch 'feature/dacol-IdData' of https://github.com/Azure/Industrial-IoT into feature/dacol-IdData

cristipogacean reviewed (approved):
:shipit:

karok2m commented on components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs:
```diff
@@ -507,6 +537,14 @@ private sealed class DataSetWriterSubscription : IDisposable {
+        /// <summary>
+        /// Helper to deserialize credential info
+        /// </summary>
+        private class cred {
```
Can we make this into a public class and use it whenever we serialize/deserialize `CredentialModel.Value` for `CredentialType.UserName` case ? I can see 3 places now where we hardcode the expected structure:
* in `PublishedNodesJobConverter.ToUserNamePasswordCredentialAsync()` method
* in `StackModelsEx.ToServiceModel()` method
* and now here
